### PR TITLE
[LM-10469] Added Universal Feed and its Styles and Customization

### PR DESCRIPTION
**Pull request by ishaan1607**
## Description

**Features**

- Added all Post Types
- Added Universal feed fragment
- Added Feed styles and flow customizations

**Bugs**

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Hotfix (fix for breaking changes)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Other, please describe:

### Author Checklist

- [x] Assignee added
- [x] Reviewer added
- [x] Description added
- [ ] Unit test cases added

### Reviewer Checklist

- [x] PR title starts with Jira ticket number prefix
- [ ] If hotfix, Add Hotfix as prefix in PR title
- [x] Label added
- [x] Commit messages follows standard format (added, changed, removed, code commented and chore)
- [x] Description about the Feature/ Bugs added

### CODE REVIEW

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/post/viewstyle/LMFeedPostViewStyle.kt [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+import com.likeminds.feed.android.core.R
+import com.likeminds.feed.android.core.ui.base.styles.*
+import com.likeminds.feed.android.core.ui.widgets.postfooterview.style.LMFeedPostFooterViewStyle
+import com.likeminds.feed.android.core.ui.widgets.postheaderview.style.LMFeedPostHeaderViewStyle
+import com.likeminds.feed.android.core.ui.widgets.postmedia.style.*
+import com.likeminds.feed.android.core.ui.widgets.posttopicsview.style.LMFeedPostTopicsViewStyle
+import com.likeminds.feed.android.core.utils.LMFeedViewStyle
+import com.likeminds.feed.android.core.utils.LMFeedViewUtils
+
+class LMFeedPostViewStyle private constructor(
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/post/viewstyle/LMFeedPostViewStyle.kt:15]: add view style description comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedIconStyle.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+import android.widget.ImageView.ScaleType
+import androidx.annotation.ColorRes
+import androidx.annotation.DrawableRes
+import androidx.core.content.ContextCompat
+import com.likeminds.feed.android.core.ui.base.views.LMFeedIcon
+import com.likeminds.feed.android.core.utils.LMFeedViewStyle
+import com.likeminds.feed.android.core.utils.model.LMFeedPadding
+
+class LMFeedIconStyle private constructor(
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedIconStyle.kt:11]: add view style description comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedIconStyle.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                    resources.getDimension(padding.paddingTop).toInt(),
+                    resources.getDimension(padding.paddingRight).toInt(),
+                    resources.getDimension(padding.paddingBottom).toInt()
+                )
+            }
+        }
+    }
+}
+
+fun LMFeedIcon.setStyle(viewStyle: LMFeedIconStyle) {
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedIconStyle.kt:101]: add comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedImageStyle.kt [hunk 4]
```diff
@@ -0,0 +1,106 @@
+package com.likeminds.feed.android.core.ui.base.styles
+
+import android.widget.ImageView.ScaleType
+import androidx.annotation.ColorRes
+import androidx.core.content.ContextCompat
+import com.likeminds.feed.android.core.R
+import com.likeminds.feed.android.core.ui.base.views.LMFeedImageView
+import com.likeminds.feed.android.core.utils.LMFeedViewStyle
+
+class LMFeedImageStyle private constructor(
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedImageStyle.kt:10]: add view style description comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedImageStyle.kt [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            if (this@LMFeedImageStyle.alpha != null) {
+                this.alpha = this@LMFeedImageStyle.alpha
+            }
+        }
+    }
+}
+
+fun LMFeedImageView.setStyle(viewStyle: LMFeedImageStyle) {
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/styles/LMFeedImageStyle.kt:104]: add comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/views/LMFeedVideoView.kt [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+import com.google.android.exoplayer2.util.Util
+import com.likeminds.feed.android.core.LMFeedCoreApplication.Companion.LOG_TAG
+import com.likeminds.feed.android.core.R
+import com.likeminds.feed.android.core.ui.widgets.postmedia.style.LMFeedPostVideoMediaViewStyle
+import com.likeminds.feed.android.core.utils.LMFeedStyleTransformer
+import com.likeminds.feed.android.core.utils.LMFeedVideoCache
+import com.likeminds.feed.android.core.utils.LMFeedViewUtils.hide
+import com.likeminds.feed.android.core.utils.LMFeedViewUtils.show
+
+class LMFeedVideoView @JvmOverloads constructor(
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/base/views/LMFeedVideoView.kt:22]: add comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/theme/LMFeedTheme.kt [hunk 7]
```diff
@@ -15,9 +19,15 @@ object LMFeedTheme {
...
+            (lmFeedSetThemeRequest.postCharacterLimit ?: DEFAULT_POST_CHARACTER_LIMIT)
     }
 
     fun getFontResources(): Pair<Int?, String?> {
         return Pair(fontResource, fontAssetsPath)
     }
+
+    fun getPostCharacterLimit(): Int {
+        return postCharacterLimit
+    }
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/theme/LMFeedTheme.kt:32]: add comments what it represent

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/widgets/headerview/styles/LMFeedHeaderViewStyle.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+import android.graphics.Typeface
+import androidx.annotation.ColorRes
+import androidx.annotation.DimenRes
+import com.likeminds.feed.android.core.R
+import com.likeminds.feed.android.core.ui.base.styles.LMFeedIconStyle
+import com.likeminds.feed.android.core.ui.base.styles.LMFeedTextStyle
+import com.likeminds.feed.android.core.utils.LMFeedViewStyle
+
+class LMFeedHeaderViewStyle private constructor(
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/widgets/headerview/styles/LMFeedHeaderViewStyle.kt:11]: add view style comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/widgets/postfooterview/view/LMFeedPostFooterView.kt [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    ) {
+
+    }
+
+    private val inflater =
+        (context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater)
+
+    private val binding = LmFeedPostFooterViewBinding.inflate(inflater, this, true)
+
+    fun setStyle(postFooterViewStyle: LMFeedPostFooterViewStyle) {
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/ui/widgets/postfooterview/view/LMFeedPostFooterView.kt:39]: add comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedFragment.kt [hunk 10]
```diff
@@ -0,0 +1,663 @@
...
+    protected open fun customizeUniversalFeedHeaderView(headerViewUniversal: LMFeedHeaderView) {
+        headerViewUniversal.apply {
+            setStyle(LMFeedStyleTransformer.universalFeedFragmentViewStyle.headerViewStyle)
+
+            setTitleText(getString(R.string.lm_feed_feed))
+        }
+    }
+
+    protected open fun onNavigationIconClick() {
+        Log.d("PUI", "default onNavigationIconClick")
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedFragment.kt:481]: remove log statements

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedFragment.kt [hunk 11]
```diff
@@ -0,0 +1,663 @@
...
+    }
+
+    protected open fun onSearchIconClick() {
+        Log.d("PUI", "default onSearchIconClick")
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedFragment.kt:485]: remove log statements

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedListView.kt [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            R.drawable.lm_feed_item_divider
+        )
+        dividerDrawable?.let { drawable ->
+            dividerDecoration.setDrawable(drawable)
+        }
+
+        addItemDecoration(dividerDecoration)
+
+        //todo: testing required
+        initiateAutoPlayer()
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedListView.kt:48]: todo , have we tested?

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedListView.kt [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        posts: List<LMFeedPostViewData>
+    ) {
+        universalFeedAdapter.replace(posts)
+    }
+
+    fun addPosts(
+        posts: List<LMFeedPostViewData>
+    ) {
+        universalFeedAdapter.addAll(posts)
+    }
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedListView.kt:107]: add comments

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/viewmodel/LMFeedUniversalFeedViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+                //convert to view data
+                val listOfPostViewData =
+                    LMFeedViewDataConvertor.convertUniversalFeedPosts(posts, usersMap, topicsMap)
+
+                //send it to ui
+                _universalFeedResponse.postValue(Pair(page, listOfPostViewData))
+            } else {
+                //for error
+//                errorMessageChannel.send(ErrorMessageEvent.UniversalFeed(response.errorMessage))
```
Comment (ishaan1607) [inline likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/viewmodel/LMFeedUniversalFeedViewModel.kt:77]: why this commented?

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedListView.kt (hunk 12 above)
Comment (sidcr7-likeminds) [reply likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedListView.kt:48]: yet to be tested, will remove this todo once testing is done

File: likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedFragment.kt (hunk 11 above)
Comment (sidcr7-likeminds) [reply likeminds-feed-android-core/src/main/java/com/likeminds/feed/android/core/universalfeed/view/LMFeedUniversalFeedFragment.kt:485]: removed in further PRs

**Pull request merged**